docs/src/main/asciidoc/hibernate-orm.adoc

```diff
@@ -542,3 +542,216 @@ Datasource configuration is extremely simple, but is covered in a different guid
 it's implemented by the Agroal connection pool extension for Quarkus.
 
 Jump over to link:datasource[Quarkus - Datasources] for all details.
+
+== Multitenancy
+
+"The term multitenancy, in general, is applied to software development to indicate an architecture in which a single running instance of an application simultaneously serves multiple clients (tenants). This is highly common in SaaS solutions. Isolating information (data, customizations, etc.) pertaining to the various tenants is a particular challenge in these systems. This includes the data owned by each tenant stored in the database" (link:https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#multitenacy[Hibernate User Guide]).
+
+Quarkus currently supports the link:https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#multitenacy-separate-database[separate database] and the link:https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#multitenacy-separate-schema[separate schema] approach. 
+
+=== Writing the application
+
+Let's start by implementing the `/{tenant}` endpoint. As you can see from the source code below it is just a regular JAX-RS resource:
+
+[source,java]
+----
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+@ApplicationScoped
+@Produces("application/json")
+@Consumes("application/json")
+@Path("/{tenant}")
+public class FruitResource {
+
+    @Inject
+    EntityManager entityManager;
+
+    @GET
+    @Path("fruits")
+    public Fruit[] getFruits() {
+        return entityManager.createNamedQuery("Fruits.findAll", Fruit.class)
+                .getResultList().toArray(new Fruit[0]);
+    }
+    
+}
+----
+
+In order to resolve the tenant from incoming requests and map it to a specific tenant configuration, you need to create an implementation for the `io.quarkus.hibernate.orm.runtime.tenant.TenantResolver` interface.
+
+[source,java]
+----
+import javax.enterprise.context.ApplicationScoped;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.Unremovable;
+import io.quarkus.hibernate.orm.runtime.tenant.TenantResolver;
+import io.vertx.ext.web.RoutingContext;
+
+@ApplicationScoped
+@Unremovable
+public class CustomTenantResolver implements TenantResolver {
+
+    @Override
+    public String getDefaultHibernateOrmTenantId() {
+        return "base";
+    }
+    
+    @Override
+    public String resolveHibernateOrmTenantId(RoutingContext context) {
```

Comment:
I think the routing context is a good choice. That's why I have taken the idea from the ODIC Multitenant implementation: [CustomTenantResolver.java](https://github.com/quarkusio/quarkus-quickstarts/blob/master/security-openid-connect-multi-tenancy/src/main/java/org/acme/quickstart/oidc/CustomTenantResolver.java).

You would normally need to get some information from the incoming request to decide which tenant is targeted with the request. The information could be taken from path (as in the example) or maybe also from the request header. 

It is also very likely that the two parts Hibernate & ODIC multtenancy will be used in combination. You can simply create a single class that shares some code between the two. Unfortunately it was not really possible to use exactly the same signature. The ODIC multtenancy relies on the fact that NULL is considered as the value for the default tenant. Hibernate does not allow a NULL value for the tenant id.

Reply:
I'm not eager to _require_ a dependency on the vertx module: it needs to be possible to use the Hibernate ORM extension even in an application which doesn't have any web endpoint.

I think there's some options:

1. Expose a different type, something like `TenantId`, or `MultitenancyContext`, and leave it to the developer to have a bean which performs the conversion - possibly making it easy such as having a CDI Factory.
2. Move your `TenantResolver` approach into a dedicated module / extension
3. Make sure that `TenantResolver` is one way out of several options - making the binding to the VertX module optional.

I suppose we can mix these approaches; for example my favourite is 1# in terms of clean design, but you could have a default factory (2#) provided for user's convenience. This could initially like in this same module (3#) and later be broken out as a future improvement.

Or maybe you have a better idea?

Reply:
OK, I think this is not a big problem. I will just rewrite the TenantResolver like this:
```java
public interface TenantResolver {
    String getDefaultTenantId();
    String resolveTenantId();
}
```
The implementation by the user then is:
```java
@RequestScoped
public class CustomTenantResolver implements TenantResolver {
    
    @Inject
    RoutingContext context;
    
    public String getDefaultTenantId() {
        return "base";
    }
    public String resolveTenantId() {
        ...
    }
}
```
So there is no longer a dependency in the Hibernate ORM extension to vertx, only in the user's implementation.

Reply:
Thanks!

The new `TenantResolver` interface looks good; question: wouldn't it be better to not have a defaultTenantId ?
If you don't force one, I would expect implementors can still choose to return a default - should other options fail. But it might also be clearer to them when the default would be used (since it's their own code), which I would find more reassuring.

Regarding the implementaiton, I do like this example but I believe it should be possible as well to not need Arc. Shouldn't the `CustomTenantResolver` rather have the option to lookup an Arc component?

Reply:
> The new TenantResolver interface looks good; question: wouldn't it be better to not have a defaultTenantId ?
> If you don't force one, I would expect implementors can still choose to return a default - should other options fail. But it might also be clearer to them when the default would be used (since it's their own code), which I would find more reassuring.

I did that first, but unfortunately Hibernate does not allow a null value for a tenantId. This means we would still have to create an artifical default value internally to use it with the Hibernate settings. Then we have to translate that back to the null value for the TenantConnectionResolver, because the user expects now also a null tenantId over there:
```java
public interface TenantConnectionResolver {
    ConnectionProvider resolve(String tenantId);
}
```
At the end I think it is easier to understand if you say "You need to define a default value for the tenantId and it cannot be null". That is what is expressed in the interface. That also matches the Hibernate philosophy which does not allow null.

Reply:
Regarding the default tenantId : ok, if you prefer so that's fine. Just making sure we're understanding the implications: don't you think it's generally a possible design error from the user's part to ever return null?
In such cases we normally prefer to throw an explict exception rather than silently and sneakily change semantics; I suspect that just requiring a tenantId would make for more robust end user systems.

Regarding the Arc puzzle:

I meant it could be the user's implementation of `TenantConnectionResolver` to explicitly have some code like 

    Arc.container().instance(RoutingContext.class);

should he/she want to base the decision on the routing - or anything else.

It's not a suggestion though, I'm not an expert of CDI nor ArC - just wondering if that's not conceptually a clean solution so to not need users to specify multiple configuration properties.